agg-ast/schema_derivation/src/schema_derivation.rs

```diff
@@ -208,6 +208,88 @@ impl DeriveSchema for Stage {
             Ok(state.result_set_schema.to_owned())
         }
 
+        fn group_derive_schema(group: &Group, state: &mut ResultSetState) -> Result<Schema> {
+            // group is a map of field namespace to expression. We can derive the schema for each expression
+            // and then union them together to get the resulting schema.
+            let mut keys = group
+                .aggregations
+                .iter()
+                .map(|(k, e)| {
+                    let field_schema = e.derive_schema(state)?;
+                    Ok((k.to_string(), field_schema))
+                })
+                .collect::<Result<BTreeMap<String, Schema>>>()?;
+            keys.insert("_id".to_string(), group.keys.derive_schema(state)?);
+            Ok(Schema::Document(Document {
+                required: keys.keys().cloned().collect(),
+                keys,
+                ..Default::default()
+            }))
+        }
+
+        // add_fields_derive_schema is near set_window_fields_derive_schema because they are necessarily
+        // identical in semantics: they both add fields to the schema incoming from the previous
+        // pipeline stage. The only difference is that add_fields is a map of expressions, while set_window_fields
+        // is a map of window functions.
+        fn add_fields_derive_schema(
+            add_fields: &LinkedHashMap<String, Expression>,
+            state: &mut ResultSetState,
+        ) -> Result<Schema> {
+            let mut keys = add_fields
+                .iter()
+                .map(|(k, e)| {
+                    let field_schema = e.derive_schema(state)?;
+                    Ok((k.to_string(), field_schema))
+                })
+                .collect::<Result<BTreeMap<_, _>>>()?;
+            let Schema::Document(Document {
+                keys: ref current_keys,
+                ..
+            }) = state.result_set_schema
+            else {
+                // This should never actually happen
+                return Err(Error::InvalidResultSetSchema(
+                    state.result_set_schema.to_string(),
+                ));
```

Comment:
Should we `panic` here instead since this is considered an unreachable state? If we got here, it would mean there is a mistake in our code, not somehow an invalid input.

Reply:
That's true, was just curious to see this if it ever happened, but it really shouldn't.